nan.h

```diff
@@ -1591,39 +1592,43 @@ class Callback {
 
 /* abstract */ class AsyncProgressWorker : public AsyncWorker {
  public:
+
+  typedef struct asyncdata_t {
+    char* data;
+    size_t size;
+    AsyncProgressWorker* worker;
+  } asyncdata_t;
+
+  typedef std::deque<asyncdata_t *> asyncqueue_t;
+
   explicit AsyncProgressWorker(Callback *callback_)
-      : AsyncWorker(callback_), asyncdata_(NULL), asyncsize_(0) {
-    async = new uv_async_t;
+      : AsyncWorker(callback_) {
     uv_async_init(
         uv_default_loop()
-      , async
+      , &async
       , AsyncProgress_
     );
-    async->data = this;
+    async.data = this;
 
     uv_mutex_init(&async_lock);
   }
 
   virtual ~AsyncProgressWorker() {
     uv_mutex_destroy(&async_lock);
-
-    if (asyncdata_) {
-      delete[] asyncdata_;
-    }
   }
 
   void WorkProgress() {
-    uv_mutex_lock(&async_lock);
-    char *data = asyncdata_;
-    size_t size = asyncsize_;
-    asyncdata_ = NULL;
-    uv_mutex_unlock(&async_lock);
-
     // Dont send progress events after we've already completed.
     if (callback) {
-        HandleProgressCallback(data, size);
+        while(!queue.empty()) {
+            uv_mutex_lock(&async_lock);
+            asyncdata_t *asyncdata = queue.front();
+            queue.pop_front();
+            uv_mutex_unlock(&async_lock);
+            HandleProgressCallback(asyncdata->data, asyncdata->size);
+            delete asyncdata;
+        }
```

Comment:
You might as well fix the indent errors in this method and Destroy() while you're here.